encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders.json;
+
+import android.util.JsonReader;
+
+import androidx.annotation.NonNull;
+
+import com.google.firebase.decoders.DataDecoder;
+import com.google.firebase.decoders.FieldRef;
+import com.google.firebase.decoders.ObjectDecoder;
+import com.google.firebase.decoders.TypeCreator;
+import com.google.firebase.decoders.TypeToken;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Map;
+//TODO: remove Exception from method header
+//TODO: surpass warnings
+//TODO: Java Docs
+public class JsonDataDecoderBuilderContext implements DataDecoder {
+  private Map<TypeToken.ClassToken<?>, ObjectDecoderContextImpl<?>> objectDecodersCtx = new HashMap<>();
+  private Map<TypeToken.ClassToken<?>, TypeCreator<?>> typeCreators = new HashMap<>();
+  private JsonReader reader;
+
+  public JsonDataDecoderBuilderContext(@NonNull Map<TypeToken.ClassToken<?>, ObjectDecoder<?>> objectDecoders) {
+    for (Map.Entry<TypeToken.ClassToken<?>, ObjectDecoder<?>> entry : objectDecoders.entrySet()) {
```

Comment:
Ideally we should make this processing only on demand, as doing this upfront may be wasteful if most registered decoders are never needed.